Adjust hearings table data displayed

beamalsky opened the pull request:
## Overview

Closes #128.
Closes #131. 

This PR responds to a list of requested changes from Jamie detailed in #128 and #131.

## Testing Instructions

 * Run the site following directions in the README
* Visit http://localhost:8000/hearings/ and make sure you see only the categories listed in `settings.DISPLAY_CATEGORIES` in the list in the top
* Click through a few pages of hearings and make sure you don't see any hearings listed with other categories
* Log in through the footer and make sure that link works
* Go back to http://localhost:8000/hearings/ and make sure you see hearings from all categories
* Visit http://localhost:8000/hearings/, a category detail page, and a committee detail page and make sure the hearings appear newest to oldest by Hearing Date and show the columns  DATE, HEARING TITLE, COMMITTEE, CATEGORY (except for COMMITTEE on committee detail pages and CATEGORY on category detail pages)

hancush commented on committeeoversightapp/models.py:
```diff
@@ -23,6 +24,25 @@
+            return HearingCategory.objects.get(
+                event_id=self.id
+            ).category
+        except ObjectDoesNotExist:
```
100% fine to import this exception for use as you've done here, just want to share some related trivia: You can access a `DoesNotExist` exception via any Django model, too, e.g., `HearingCategory.DoesNotExist`. Accomplishes the same thing, just minimally more convenient. 🙃

hancush commented on committeeoversightapp/views.py:
```diff
@@ -135,46 +171,28 @@ def prepare_results(self, qs):
-            eventparticipant__event=context['hearing'].id
-        )
-
+        context['category'] = context['hearing'].category
```
What's the reason for explicitly setting these in the context, instead of accessing them via your `hearing` object in the template?

beamalsky replied:
A vestigial structure. Ty for catching!

hancush commented on committeeoversightapp/views.py:
```diff
@@ -111,18 +121,44 @@ def prepare_results(self, qs):
-                        escape(item.name.title())
-                    ),
-                    item.start_date,
+        for item in qs:
```
Hm, the logic in this loop is getting pretty tough to follow. Can you think of a way to break this into methods or roll some of this conditional string formatting back up into model properties?

beamalsky replied:
Refactored! Thanks for the nudge.

hancush commented on committeeoversightapp/views.py:
```diff
@@ -74,10 +75,19 @@ class EventList(LoginRequiredMixin, TemplateView):
+    columns = [
+        'start_date',
+        'name',
+        'id', # committees
```
Are you required to include a value here, or can it be an empty string?

beamalsky replied:
This does work with an empty string. Good idea!

hancush commented on committeeoversightapp/views.py:
```diff
@@ -135,46 +171,28 @@ def prepare_results(self, qs):
+                ]
+                json_data.append(row_data)
+            # if not authenticated, only include display hearing categories 
+            elif not item.category or item.category.name in settings.DISPLAY_CATEGORIES:
```
This is a little confusing to me. If there is not an item category, or if the category is in display categories, add two empty values?

hancush replied:
`category_string` will be an empty value if there is not a category, and it is added to the row unconditionally in L156. Will this introduce a redundant empty value?

beamalsky replied:
I think my refactor addresses this question. I also realized that I'm not actually sure if hearings without a category should be displayed to non-admin users. I opened a question issue for it (#134) and emailed Jamie. For now, hearings with no category will only show up for admin users.

beamalsky commented:
The hearing list view logic should be easier to follow now. Ready for your eyes Monday @hancush !

hancush commented on committeeoversightapp/templates/partials/hearing_table.html:
```diff
@@ -2,9 +2,10 @@
-         <th scope="col">Date Updated</th>
-         <th scope="col" class="hearing-title-col">Hearing Title</th>
-         <th scope="col">Hearing Date</th>
+         <th scope="col" style="min-width:10%;">Date</th>
```
TIL you should do this, instead of the `col-*` classes, in Bootstrap 4! Thank you!!!

beamalsky commented:
Thx for the catch @hancush! The error you were seeing was from an incorrect column value in the committee detail page DataTables settings, which I've fixed. 

In my last commit I also caught an error in how I was building the JSON for the server side processing. I noticed that for non-authenticated users, reordering the columns after the page initially loaded caused the table to have some disappearing rows, showing 2 or 3 entries instead of 10. I figured out that this is because I was doing my filtering for authenticated users in the `prepare_results` function, which it seems expects an output of the same length as the input `qs`. I moved this filtering to the `filter_queryset` function in that view, where it should have been all along.
